scripts/sct_run_batch.py

```diff
@@ -21,146 +21,365 @@
+
+    if args.exclude is not None:
+        subject_dirs = [f for f in subject_dirs if re.search(args.exclude, f) is None]
+
+    if args.exclude_list is not None:
+        subject_dirs = [f for f in subject_dirs if f not in args.exclude_list]
+
+    # Determine the number of jobs we can run simulataneously
+    if args.jobs < 1:
+        jobs = multiprocessing.cpu_count() + args.jobs
+    else:
+        jobs = args.jobs
+
+    # Run the jobs, recording start and end times
+    start = time.strftime('%H:%M', time.localtime(time.time()))
+
+    # Trap errors to send an email if a task fails.
+    try:
+        with multiprocessing.Pool(jobs) as p:
+            run_single_dir = functools.partial(run_single,
+                                               task=task,
+                                               task_args=args.task_args,
+                                               path_segmanual=path_segmanual,
+                                               path_data=path_data,
+                                               path_results=path_results,
+                                               path_log=path_log,
+                                               path_qc=path_qc,
+                                               itk_threads=args.itk_threads,
+                                               continue_on_error=args.continue_on_error)
+            results = p.map(run_single_dir, subject_dirs)
+            end = time.strftime('%H:%M', time.localtime(time.time()))
+    except Exception as e:
+        if do_email is not None:
+            message = ('Oh no there has been the following error in your pipeline:\n\n'
+                       '{}'.format(e))
+            try:
+                # I consider the multiprocessing error more significant than a potential email error, this
+                # ensures that the multiprocessing error is signalled.
+                send_notification('sct_run_batch errored', message)
+            except Exception:
+                raise e
+
+            raise e
+        else:
+            raise e
+
+    end = time.strftime('%H:%M', time.localtime(time.time()))
+
+    # Check for failed subjects
+    fails = [sd for (sd, ret) in zip(subject_dirs, results) if ret.returncode != 1]
+
+    completed_message = ('Finished ', ':-)\n' if len(fails) == 0 else '\n'
+                         'Started: {}\n'
+                         'Ended: {}\n'
+                         ''.format(start, end))
+
+    if len(fails) == 0:
+        status_message = 'Hooray your batch completed successfully'
+    else:
+        status_message = ('Your batch completed but some subjects may have not completed '
+                         'successfully, please consult the logs for:\n'
+                         '{}\n'.format('\n'.join(fails)))
+
+    print(status_message + completed_message)
+
+    if do_email:
+        send_notification('sct_run_batch: Run completed',
+                          fails_message + completed_message)
```

Comment:
`fails_message` is not declared. Did you mean `fails`?

Reply:
whoops, that should be `status_message`, fixed